larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,623 @@
+		      {
+		          nearestHit = distance;
+
+		      }
+		      
+		      if (std::abs( vector1.GetDotProduct(centroidVector) ) > smallestAngle)
+		         continue;
+                    
+                      if (caloHit1->GetPseudoLayer() == caloHit2->GetPseudoLayer())
+	              {
+			      if ( distance < m_proximityThreshold)
+			      { 
+				      ++proximity;
+				      if ( distance < m_contactThreshold)
+				      {
+					      ++contact;
+					      --proximity;
+				      }
+		               }
+
+	              }
+		      smallestAngle = std::abs( vector1.GetDotProduct(centroidVector));
+		      finalDistance = distance;
+                  }
+		  distanceDistribution.push_back(finalDistance);
+	    }
+
+            float combinedNClusterHits(clusterNHits + otherClusterNHits);
+	    float combinedNClusterEdgeHits(clusterNEdgeHits + otherClusterNEdgeHits);
+ 
+            const double avgDistance{!distanceDistribution.empty() ? std::accumulate(distanceDistribution.begin(), distanceDistribution.end(), 0.0) / distanceDistribution.size() : -1.0};
+
+	    std::vector<std::string> featureOrder;
+	    const bool mcMatch(clusterMC == otherClusterMC);
+
+	    featureOrder.emplace_back("VertexClusterAngle");
+            featureOrder.emplace_back("MinEdgeHitSeparation");
+            featureOrder.emplace_back("Cluster1NHits");
+            featureOrder.emplace_back("Cluster2NHits");
+            featureOrder.emplace_back("NHitsInContact");
+            featureOrder.emplace_back("NHitsInProximity");
+            featureOrder.emplace_back("CentroidSeparation");
+            featureOrder.emplace_back("AvgDistance");
+	    featureOrder.emplace_back("CentroidVectorX");
+	    featureOrder.emplace_back("CentroidVectorZ");
+	    featureOrder.emplace_back("Cluster1ADC");
+	    featureOrder.emplace_back("Cluster2ADC");
+
+	    if (m_trainingSetMode)
+            {
+                            
+                LArMvaHelper::MvaFeatureVector featureVector;
+                
+		featureVector.emplace_back(static_cast<double>(vtxClusterAngle));
+		featureVector.emplace_back(static_cast<double>(minEdgeHitSeparation));
+                featureVector.emplace_back(static_cast<double>(clusterNHits));
+                featureVector.emplace_back(static_cast<double>(otherClusterNHits));
+                featureVector.emplace_back(static_cast<double>(contact));
+		featureVector.emplace_back(static_cast<double>(proximity));
+ 		featureVector.emplace_back(static_cast<double>(centroidSeparation));
+		featureVector.emplace_back(static_cast<double>(avgDistance));
+		featureVector.emplace_back(static_cast<double>(centroidVector.GetX()));
+		featureVector.emplace_back(static_cast<double>(centroidVector.GetZ()));
+		featureVector.emplace_back(static_cast<double>(adc1));
+		featureVector.emplace_back(static_cast<double>(adc2));
+
+                LArMvaHelper::ProduceTrainingExample(m_trainingOutputFile, mcMatch, featureVector);
```

Comment:
I think this answers my question above about whether the MC is needed.  Still, it feels a little odd to always extract MC information when running reconstruction.  If its possible to factorise out the needed MC into a separate function, I think that should happen.  Otherwise leave as is.